### ISSUE TITLE
walrus-assignment inference differs from single-assignment inference

### ISSUE THREAD

**Issue by Andrej730** [microsoft/pyright#8054]
I've noticed the following behaviour - using walrus assignment is treated the same way as if there wasn't an assignment. See example code below.
Shouldn't it be treated the same way as a single-assignment? Typically walrus operator is used to just reduce two lines to one but intended meaning is the same.

Also noticed an issue with tooltip in this case (I use Pylance v2024.5.103 (pre-release)), it shows two contradicting lines:
![image](https://github.com/microsoft/pyright/assets/9417531/6c7879b5-a2b0-4208-9163-a8a0c0f681b5)

```python
class A: ...
class B(A): ...

def func(o: list[A]) -> None: ...

# single-assignment inference results in type list[B],
# therefore a type error
a1 = [B()]
func(a1) # fail

# bidirectional inference results in type list[A]
# therefore it's allowed
a2 = B()
func([a2]) # allowed

# bidirectional inference takes precedence over walrus-assignment
# and type results in list[A], therefore it's allowed
func(a3:=[B()]) # allowed, shouldn't it be list[B]
reveal_type(a3) # Type of "a3" is "list[A]"
```

**Comment by erictraut** [comment]
Pyright is working as intended here. In the case where you use the walrus operator, bidirectional type inference can be used because there is an "expected type" provided by the `func` call. This influences the evaluated type of the list expression `[(B)]`. 

You can see that `func([B()])` is type safe because `[B()]` evaluates to `list[A]` thanks to bidirectional type inference. When you add `a3 :=` to this expression, it simply assigns this evaluated type to `a3`.

**Comment by Andrej730** [comment]
I understand, it just contradicts with what I would expect thinking that any assignment including walrus operator will be more strict and will use single-assignment inference. Just a note that it's different from mypy.

Any ideas about the issue in the tooltip (it says that `a3` is both `list[A]` and `list[B]`)? Or is it Pylance issue?

**Comment by erictraut** [comment]
Mypy's behavior here is demonstrably wrong. It produces no error for the expression `func([B()])` (which is correct), but it produces an error for the expression `func(a3 := [B()])`. The two expression are calling the same function with the same value, so neither of them should result in a type violation. The only difference between the two is that the second expression is assigning an intermediate value to a variable `a3`.

I missed your point about the tooltip in the earlier post. This is indicative of a bug in pyright, so I'm going to reopen and investigate that further.

**Comment by erictraut** [comment]
The bug that causes the apparent inconsistency in the tooltip will be fixed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.366.
